src/stubadub/core.cljc

```diff
@@ -1,29 +1,140 @@
+       (-invoke [this x1 x2 x3 x4 x5 x6 x7 x8 x9 x10 x11]
+                (invoke-stub this [x1 x2 x3 x4 x5 x6 x7 x8 x9 x10 x11]))
+       (-invoke [this x1 x2 x3 x4 x5 x6 x7 x8 x9 x10 x11 x12]
+                (invoke-stub this [x1 x2 x3 x4 x5 x6 x7 x8 x9 x10 x11 x12]))
+       (-invoke [this x1 x2 x3 x4 x5 x6 x7 x8 x9 x10 x11 x12 x13]
+                (invoke-stub this [x1 x2 x3 x4 x5 x6 x7 x8 x9 x10 x11 x12 x13]))
+       (-invoke [this x1 x2 x3 x4 x5 x6 x7 x8 x9 x10 x11 x12 x13 x14]
+                (invoke-stub this [x1 x2 x3 x4 x5 x6 x7 x8 x9 x10 x11 x12 x13 x14]))
+       (-invoke [this x1 x2 x3 x4 x5 x6 x7 x8 x9 x10 x11 x12 x13 x14 x15]
+                (invoke-stub this [x1 x2 x3 x4 x5 x6 x7 x8 x9 x10 x11 x12 x13 x14 x15]))
+       (-invoke [this x1 x2 x3 x4 x5 x6 x7 x8 x9 x10 x11 x12 x13 x14 x15 x16]
+                (invoke-stub this [x1 x2 x3 x4 x5 x6 x7 x8 x9 x10 x11 x12 x13 x14 x15 x16]))
+       (-invoke [this x1 x2 x3 x4 x5 x6 x7 x8 x9 x10 x11 x12 x13 x14 x15 x16 x17]
+                (invoke-stub this [x1 x2 x3 x4 x5 x6 x7 x8 x9 x10 x11 x12 x13 x14 x15 x16 x17]))
+       (-invoke [this x1 x2 x3 x4 x5 x6 x7 x8 x9 x10 x11 x12 x13 x14 x15 x16 x17 x18]
+                (invoke-stub this [x1 x2 x3 x4 x5 x6 x7 x8 x9 x10 x11 x12 x13 x14 x15 x16 x17 x18]))
+       (-invoke [this x1 x2 x3 x4 x5 x6 x7 x8 x9 x10 x11 x12 x13 x14 x15 x16 x17 x18 x19]
+                (invoke-stub this [x1 x2 x3 x4 x5 x6 x7 x8 x9 x10 x11 x12 x13 x14 x15 x16 x17 x18 x19]))
+       (-invoke [this x1 x2 x3 x4 x5 x6 x7 x8 x9 x10 x11 x12 x13 x14 x15 x16 x17 x18 x19 x20]
+                (invoke-stub this [x1 x2 x3 x4 x5 x6 x7 x8 x9 x10 x11 x12 x13 x14 x15 x16 x17 x18 x19 x20]))
+       (-invoke [this x1 x2 x3 x4 x5 x6 x7 x8 x9 x10 x11 x12 x13 x14 x15 x16 x17 x18 x19 x20 more]
+                (apply invoke-stub this x1 x2 x3 x4 x5 x6 x7 x8 x9 x10 x11 x12 x13 x14 x15 x16 x17 x18 x19 x20 more))]))
+
+(defn- parse-opts-and-body [opts-and-body]
   (let [has-return-value? (= :returns (first opts-and-body))
         has-return-fn? (= :return-fn (first opts-and-body))
-        has-opts? (or has-return-value? has-return-fn?)
-        body (if has-opts?
-               (nnext opts-and-body)
-               opts-and-body)
-        return-value (when has-opts? (second opts-and-body))
-        args (gensym)]
-    `(with-redefs [*calls* (if @*calls*
-                             *calls* (atom []))
-                   ~func (fn [& ~args]
-                           (swap! *calls*
-                                  conj {:func ~func, :args ~args})
-                           ~(cond
-                              has-return-value? return-value
-                              has-return-fn? `(~return-value ~args)
-                              :else nil))]
+        has-opts? (or has-return-value? has-return-fn?)]
+    {:opts {:return-value (when has-opts? (second opts-and-body))
+            :has-return-value? has-return-value?
+            :has-return-fn? has-return-fn?}
```

Comment:
What do you think about simplifying this so that `:opts` either has `:return-value` or `:return-fn`? Instead of overloading `:return-value` and then checking with flags?

Reply:
I think that is a great idea. Implementation was left untouched as far as I could.

I'll fix.

Reply:
Fixed and refactored. The base mock construct now only takes a function.